pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,294 @@
+				return acadmission.Allowed("audit policy did not change")
+			}
+		}
+
+		// Validate audit policy semantically
+		auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+		}
+		auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+		if !ok {
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+		}
+		errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+		if len(errList) != 0 {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+		}
+
+		// Validate schema version against Shoot Kubernetes version
+		existingShoots := &gardencorev1beta1.ShootList{}
+		if err := h.apiReader.List(ctx, existingShoots, client.InNamespace(request.Namespace)); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if err := validatePolicyAgainstShoots(request, schemaVersion, existingShoots); err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		return acadmission.Allowed("configmap change is valid")
+	}
+	return acadmission.Allowed("operation is not update or delete")
+}
+
+// IsValidAuditPolicyVersion checks whether the api server support the provided audit policy apiVersion
+func IsValidAuditPolicyVersion(shootVersion string, schemaVersion *schema.GroupVersionKind) (bool, error) {
+	auditGroupVersion := schemaVersion.GroupVersion().String()
+
+	if auditGroupVersion == "audit.k8s.io/v1" {
+		return version.CheckVersionMeetsConstraint(shootVersion, ">= v1.12")
+	}
+	return true, nil
+}
+
+func (h *handler) getOldObject(request admission.Request, oldObj runtime.Object) error {
+	if len(request.OldObject.Raw) != 0 {
+		if err := h.decoder.DecodeRaw(request.OldObject, oldObj); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("could not find old object")
+	}
+	return nil
+}
+
+func getAuditPolicy(cm *corev1.ConfigMap) (auditPolicy string, err error) {
+	auditPolicy, ok := cm.Data[auditPolicyConfigMapDataKey]
+	if !ok {
+		return "", fmt.Errorf("missing '.data.policy' in audit policy configmap")
+	}
+	if len(auditPolicy) == 0 {
+		return "", fmt.Errorf("empty audit policy. Provide non-empty audit policy")
+	}
+	return auditPolicy, nil
+}
+
+func validatePolicyAgainstShoots(request admission.Request, schemaVersion *schema.GroupVersionKind, shoots *gardencorev1beta1.ShootList) error {
+	for _, s := range shoots.Items {
+		if hasAuditPolicy(&s) {
+			cmRef := s.Spec.Kubernetes.KubeAPIServer.AuditConfig.AuditPolicy.ConfigMapRef
+			if cmRef.Name == request.Name {
+				if isValidVersion, err := IsValidAuditPolicyVersion(s.Spec.Kubernetes.Version, schemaVersion); err != nil {
+					return err
+				} else if !isValidVersion {
```

Comment:
I guess you don't need `else if` here, or? Just another `if` case would do the same and reduces nesting (= improves readability)?

Reply:
It is a bit less succinct as I need to do the `err!= nil` check in a separate line. But I agree it is more easy to read.